usr/src/lib/libzfs/common/libzfs_mount.c

```diff
@@ -1414,6 +1434,14 @@ zfs_foreach_mountpoint(libzfs_handle_t *hdl, zfs_handle_t **handles,
 	 */
 	for (int i = 0; i < num_handles;
 	    i = non_descendant_idx(handles, num_handles, i)) {
+		/*
+		 * Since the mountpoints have been sorted so that the zoned
+		 * filesystems are at the end, a zoned filesystem seen from
+		 * the global zone means that we're done.
+		 */
+		if (zoneid == GLOBAL_ZONEID &&
+		    zfs_prop_get_int(handles[i], ZFS_PROP_ZONED))
+			break;
```

Comment:
This loop only iterates over top-level mountpoints to kick off the top-level mount tasks. Isn't it possible that the sorting will result in this loop never hitting any of the zoned filesystems, and instead their being found by one of the zfs_mount_task() tasks that were spawned by this loop?

For example, if you have:

/a
/b
/b/1 (zone)

Won't the task spawned for `/b` result in the mounting of `/b/1`?

Reply:
Thanks for looking.

Yes, in that case the `/b` task will attempt to mount `/b/1` but that will fail at https://github.com/openzfs/openzfs/blob/master/usr/src/lib/libzfs/common/libzfs_mount.c#L353 ; that is no worse than the existing code which actually attempts to mount all zoned filesystems.

The fix for the race would work just as well without this test and break, but this is an optimisation that reduces the number of spawned threads in most cases, and that reduction can be significant so I thought it worthwhile.

Reply:
Got it, thanks for the clarification, the fix looks good.